### PR TITLE
Implement user authentication

### PR DESCRIPTION
**Pull request by samueltangz**

### CODE REVIEW

File: back-end/src/db.js [hunk 1]
```diff
@@ -3,23 +3,24 @@ const mongoose = require('mongoose')
...
 const Users = require('./models/users')
 
 mongoose.connection.on('error', function (err) {
   console.error('could not connect to the database')
   console.error(err)
   process.exit()
 })
-mongoose.connection.once('open', function () {
-  console.log('connected to the database')
-})
```
Comment (harrier-lcc) [inline back-end/src/db.js]: Why remove this connecting log?

File: back-end/src/api/admin/challenges.js [hunk 2]
```diff
@@ -1,60 +1,55 @@
...
+  })
 })
 
 // GET /challenges - List the challenges
 router.get('/', async function (req, res) {
   const challenges = await db.Challenges.findWithAuthor()
   return res.status(200).json({
     challenges: challenges.map(marshalChallenge(verboseLevels.SIMPLE))
-  }).end()
+  })
```
Comment (harrier-lcc) [inline back-end/src/api/admin/challenges.js]: :o API changed without closing end?

File: back-end/src/db.js (hunk 1 above)
Comment (samueltangz) [reply back-end/src/db.js]: Moved to `back-end/src/index.js` as it is too verbose to see the logs in the tests.

File: back-end/src/api/admin/challenges.js [hunk 3]
```diff
@@ -1,60 +1,55 @@
...
-], async function (req, res) {
-  const errors = validationResult(req)
-  if (!errors.isEmpty()) {
-    return res.status(400).json({
-      errors: errors.array()
-    }).end()
-  }
+], validateInputs, async function (req, res) {
   const id = parseInt(req.params.id, 10)
   const challenge = await db.Challenges.findOneWithAuthor(id)
```
Comment (harrier-lcc) [inline back-end/src/api/admin/challenges.js]: :/ isnt this finding authorid, not challenge id?
I notice the Challenge.statics.findOneWithAuthor match followed by lookup logic, if this does both, I think its better to seperate those two as two API.

File: back-end/src/api/challenges.js [hunk 4]
```diff
@@ -11,39 +12,35 @@ router.get('/', async function (req, res) {
...
-], async function (req, res) {
-  const errors = validationResult(req)
-  if (!errors.isEmpty()) {
-    return res.status(400).json({
-      errors: errors.array()
-    }).end()
-  }
+], validateInputs, async function (req, res) {
   const id = parseInt(req.params.id, 10)
   const challenge = await db.Challenges.findOneWithAuthor(id)
```
Comment (harrier-lcc) [inline back-end/src/api/challenges.js]: same issue with admin/challenge

File: back-end/src/api/admin/challenges.js (hunk 3 above)
Comment (samueltangz) [reply back-end/src/api/admin/challenges.js]: according to [L52 of `models/challenges.js`](https://github.com/black-bauhinia/pwnable.hk/blob/implement-user-authentication/back-end/src/models/challenges.js#L52), `this` is actually the challenge model.

File: back-end/src/api/challenges.js (hunk 4 above)
Comment (samueltangz) [reply back-end/src/api/challenges.js]: same

File: back-end/src/helpers/middlewares.js [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+  const { user } = req
+  if (getRoleLevel(user.role) < getRoleLevel(role)) {
+    return res.status(403).json({ error: 'you do not have enough permissions' })
+  }
+  next()
+}
+
+module.exports.validateInputs = validateInputs
+module.exports.authenticate = authenticate
+module.exports.authorize = authorize
```
Comment (harrier-lcc) [inline back-end/src/helpers/middlewares.js]: why not 
```
module.exports = {
  validateInputs,
  ...,
}
```
? Or am i misunderstood js sosad

Comment (samueltangz) [reply back-end/src/helpers/middlewares.js]: You are right... though they are equivalent

File: back-end/src/models/counters.js [hunk 6]
```diff
@@ -0,0 +1,26 @@
...
+    type: String,
+    unique: true
+  },
+  count: {
+    type: Number,
+    default: 0
+  }
+})
+
+Counter.statics.increaseAndGetCount = async function (name) {
```
Comment (harrier-lcc) [inline back-end/src/models/counters.js]: Why add an specific counter? @@
Couldnt this be done in the database level (through atomic id etc?)

File: back-end/src/helpers/middlewares.js (hunk 5 above)
Comment (harrier-lcc) [reply back-end/src/helpers/middlewares.js]: Yes, but this is cleaner lol (for unity as well, though i am fine if its one line only)

File: back-end/src/models/counters.js (hunk 6 above)
Comment (samueltangz) [reply back-end/src/models/counters.js]: https://www.tutorialspoint.com/mongodb/mongodb_autoincrement_sequence

I admit that my decision on using mongo is absolutely wrong. :)

File: back-end/src/api/admin/challenges.js (hunk 3 above)
Comment (harrier-lcc) [reply back-end/src/api/admin/challenges.js]: ? i know `this` is the challenge model......
Can you clearify what `findOneWithAuthor` actually do?
Is it 1) Find a challenge with authorid; 2) Find a challenge with given id; or 3) Both?

File: back-end/src/helpers/middlewares.js (hunk 5 above)
Comment (samueltangz) [reply back-end/src/helpers/middlewares.js]: Fixed anyway

File: back-end/src/api/admin/challenges.js (hunk 3 above)
Comment (samueltangz) [reply back-end/src/api/admin/challenges.js]: `Challengs.findOneWithAuthor(id)` is to find a `challenge`, along with the correspond author, by the given (challenge) id.
Theoretically it is equivalent to:
```SQL
SELECT challenges.*, user.username
  FROM challenges
  LEFT JOIN users ON challenges.authorId = users.id
```

File: back-end/src/api/users.js [hunk 7]
```diff
@@ -1,37 +1,58 @@
...
+    password: { salt, verifier }
+  })
+
+  return res.status(201).end()
 })
 
 // GET /users - List users
 router.get('/', async function (req, res) {
+  // TODO
   return res.status(200).json({
```
Comment (harrier-lcc) [inline back-end/src/api/users.js]: be consistent, use 405?

Comment (samueltangz) [reply back-end/src/api/users.js]: I guess I'll leave this one undone as the API should be implemented very soon.

**Pull request merged**